README.txt
BSDS 100 Final Project

Chezka Sino
Sean Tey

final-project/news.txt
Alphabet claims proof of collusion between engineer, Uber,Alphabet: Winning Through Innovation,Why Alphabet Inc (GOOGL) Stock Will Blow Past $1000,Google parent Alphabet&#39;s profit up 29 percent on strong ad sales,How the Alphabet Inc (GOOGL) Google Docs Attack Helps Microsoft,Alphabet Inc. Earnings Climb 29% In Q1,Are Alphabet&#39;s Other Bets Paying Off?,Alphabet Inc (NASDAQ:GOOGL): YouTube Betting on New Original Shows,Alphabet Inc. (GOOGL) Has Surged To A New High On Strong Q1 Earnings,Alphabet Inc Is Bringing Several New Original Series To YouTube,Alphabet Inc Stock Soars To Record High Ahead Of Earnings,Will Alphabet Inc (GOOG, GOOGL) Whiff on Earnings Yet Again?,Accounting Adjustments Make Alphabet&#39;s Valuation Below Market And Too Pessimistic,Alphabet Inc (NASDAQ:GOOGL) Threatens Microsoft Corporation (NASDAQ:MSFT) in ...,Alphabet: Coming Price Range As Seen By Big$ Funds,Will Google Whiff on Earnings Yet Again?,Capital Advantage, Inc. Buys Clorox Co, Dow Chemical Co, Alphabet Inc, Sells ...,Alphabet Inc (NASDAQ:GOOGL) Subsidiary Waymo Puts Pressure on Uber for Stolen IP,The Wonderful Secret Behind Options Earnings Trading in Alphabet Inc (NASDAQ ...,Earnings Previews: Amazon.com, Inc., Alphabet Inc, Microsoft Corporation,Alphabet Inc (NASDAQ:GOOGL): Schmidt Calls Proposed H-1B Visas Limits ...,Alphabet Offers A Glimpse Of The Self-Driving Future,Why Alphabet Inc (GOOGL) Stock Is STILL Easy Money,Earnings Preview: What To Expect From Google On Thursday,Alphabet Inc (GOOGL) Gives Investors Reason To Smile, Expedia Inc (EXPE) Stock ...,Alphabet Inc (NASDAQ:GOOGL) Acquires Land Situated Next to Tesla Gigafactory,Google: The Most Valuable Company In The World,Google&#39;s Verily Says, &#39;We Grew Up.&#39; We&#39;ll See,Alphabet Inc Is Creating A New Job Platform, Dubbed Google Hire,Buy Alphabet Inc (GOOGL) Stock With 16% Off,Price Alert: Alphabet Inc (NASDAQ:GOOGL) Stock Nears Resistance -- Technicals ...,Will Alphabet Inc New Highs Result In Another Price Collapse?,Alphabet Inc&#39;s (GOOGL) Google Is in a Tough Spot,RBC Capital&#39;s Mark Mahaney Offers Q1 Earnings Run-Up on Amazon.com, Inc. (AMZN ...,Quick Preview Of Google&#39;s Earnings Report,Nvidia Corporation (NVDA), Alphabet Inc (GOOGL), Twitter Inc (TWTR) - Today&#39;s ...,Pictet Asset Management Ltd Buys Alphabet Inc, Qualcomm Inc, Kinder Morgan Inc ...,Alphabet Inc. (GOOG) sitting close to a major resistance line on the technical ...,Congress Asset Management Co. MA Acquires New Position in Alphabet Inc (GOOGL),Alphabet Inc Reaches 1 Year High of $920.77,Alphabet Inc (GOOGL): What are the future prospects?,Alphabet Inc (GOOGL): Strong Q1 Results from A to Z,Technical Analysis for today&#39;s market share : Alphabet Inc. (NASDAQ:GOOG),Alphabet Inc Reaches 1 Year High of $942.99,Google earnings: A new approach arrives amid YouTube ad controversy,Boulegeris Investments, Inc. Buys Alphabet Inc, Paychex Inc, Bristol-Myers ...,Trading Options During Earnings in Alphabet Inc (NASDAQ:GOOGL),Google servers launch in Cuba,Alphabet Inc (GOOGL) Drive Core Revenues Faster Than Expected,Google-parent Alphabet applies to turn 12-acres of Toronto&#39;s waterfront into ...,Alphabet Inc. (GOOG) And Its Technical Chart Creates A Compelling Picture,3 Big Stock Charts for Friday: Alphabet Inc (GOOGL), Intel Corporation (INTC ...,The Android Settlement In Russia By Yandex NV (NASDAQ:YNDX), Alphabet Inc ...,The $2 trillion earnings day: Comcast and Abbvie gain, Alphabet and Amazon on deck,Alibaba Group Holding Limited and Alphabet Inc. Head to Head Compare,Alphabet Inc (GOOGL) Will Report Q1&#39;17 Results on April 27,Morning Recap: Alphabet Inc. (NASDAQ: GOOG),Buy or Sell? What Analysts Recommends: Alphabet Inc. (GOOG), The Hartford ...,Taking a Fresh Look at Alphabet Inc. (GOOGL),Alphabet Inc (NASDAQ:GOOGL) Introduces &#39;Similar Items&#39; to Enable Shopping ...,Alphabet Inc (NASDAQ:GOOGL) Offers to Invest in LG Display to Secure OLED Supply,Technical Analysis Perspective: Alphabet Inc. (GOOG), APPAREL HOLDING CORP. (VNCE),Alphabet Inc. Stock Is A Temptation To Resist,Alphabet Inc. (GOOG), AmerisourceBergen Corporation (ABC) Stock Eyes Ahead of ...,YouTube to produce more than 40 original shows,Baskin Financial Services Inc. Buys Live Nation Entertainment Inc, Alphabet ...,Alphabet Inc (GOOGL) Rating Reiterated by BMO Capital Markets,Stocks: Alphabet Inc (GOOGL) Earns “Overweight” Rating from Cantor Fitzgerald,Stocks: Alphabet Inc (GOOG) Given a $995.00 Price Target at Macquarie,Google says introducing Next Generation Jump Camera— The Yi Halo—and Jump Start,Google says one Google Home to support up to six users- Blog,Google introduces new version of Google Earth on web and Android,Google says android pay integrating with mobile banking apps - blog,Google says there is no gender pay gap at the company - Blog,Google says Fact Check now available in Google Search and News around the world,Neustar Inc announced new advanced marketing analytics partnership with Google,Google Selects comScore as Brand Safety Measurement Provider,Google says announcing patent licensing initiative, PAX,Google says bringing Google Calendar to the iPad,Google says users can now connect Google Assistant to Best Buy Insignia, LIFX, TP-Link, Wink,Moody&#39;s says Alphabet&#39;s loss of &quot;notable advertisers&quot; on YouTube platform credit negative, but does not impact AA2 rating,Ride-Sharing service Lyft latest to remove Youtube ads - NYT,Google says soon rolling out location sharing on Google maps worldwide,Google in Europe says begun a thorough review of ads policies and brand controls,Google says introducing feature to share money through Gmail app on android,VideoBurst starts cooperation with Google,Movado Group announces a partnership with Google to launch Movado Connect,Novartis Chairman: No decision yet on Alcon&#39;s future,Google says users can pre-register for games coming to Play Store,Comcast to launch YouTube on Xfinity X1,S&amp;P upgrades Alphab

final-project/server.R
#
# This is the server logic of a Shiny web application. You can run the 
# application by clicking 'Run App' above.
#
# Find out more about building applications with Shiny here:
# 
#    http://shiny.rstudio.com/
#

library(shiny)
library(ggplot2)
library(fImport)
library(quantmod)
library("tm")
library("SnowballC")
library("wordcloud")
library("RColorBrewer")
require(XML)
require(plyr)
require(stringr)
require(lubridate)

max_plots <- 10

shinyServer(function(input, output) {
  
  formulaText <- reactive({
    input$text
  })
  
  
  output$caption <- renderText({
    formulaText()
    # input$variable
  })
  
  stocks <- reactiveValues()

  # observeEvent(input$add, {
  #   stocks$dList <- c(isolate(stocks$dList), input$text)
  # })
  
  output$list <- renderText({
    stocks$dList
  })
  
  output$metrics <- renderTable({
    

    getMetrics(input$text)
  })
  
  output$stockPlot <- renderPlot({
    
    withProgress(message = 'Loading', value = 0, {

      
    incProgress(2/10, detail = paste("Loading Stock Data from API"))
        

   
    
    #stockData <- createStockData(input$text)

    inputStockName <- paste(input$text, ".", input$variable, sep="")

    # Note! input$variable needs the format [stockName.varaible] e.g. "GOOG.Adj.Close"
    # So I suggest concat or paste stockName+input$variable

    #stockData <- stockData[which(as.Date(row.names(stockData), "%Y-%m-%d") > max(as.Date(row.names(stockData), "%Y-%m-%d")) - 
    #                               as.numeric(input$select)),]
    
    #data <- data.frame(date = row.names(stockData), var = stockData[[inputStockName]])
    y.lab <- paste(input$variable, "Price", sep=" ")
    x.lab <- ""
    
    # Harcoded 2017 x-axis label
    if (input$select == "9" | input$select == "32" | input$select == "92") {
      x.lab <- 2017
    }
    
    if (input$variable == "Volume") {
      y.lab <- "Volume"
    }
    
    incProgress(4/10, detail = paste("Plotting Data.."))
    #ggplot(data, aes(as.Date(date, "%Y-%m-%d"), var)) + geom_line() + labs(y = y.lab, x = x.lab)
    
    
    if (length(input$TA)!=0){
      indicators <- paste0(paste0(input$TA,collapse = ""),"addVo()")
    } else {
      indicators <- "addVo()"
    }
    
    StockData <- createStockData(input$text)
    chartSeries(StockData,theme = "black",subset = input$select,TA=indicators)    
    
    
    })# With Progress Function End
    
  })
  
  output$wordCloud <- renderPlot({
    
    withProgress(message = 'Loading', value = 0, {
      
    filepath <- "news.txt"
    
    incProgress(5/10, detail = paste("Downloading News Information"))
    
    createNewsTXT(input$text, 100, filepath)
    # Read the text file from internet
    # filePath <- "http://www.sthda.com/sthda/RDoc/example-files/martin-luther-king-i-have-a-dream-speech.txt"
    # text <- readLines(filePath)
    
    text <- readLines(filepath)
    
    incProgress(8/10, detail = paste("Drawing Word Cloud"))
    drawWordCloud(text)
    
    
    
    })# With Progress Function End
    
    ## Make word cloud available draw space bigger here
    } , height = 750, width = 1000
  
  )
  
  output$newsTitle <- renderText({
    withProgress(message = 'Loading', value = 0, {
    incProgress(90/100, detail = paste("Loading News Data"))
    title <- paste("Top news for ", input$text, ":", sep="")
    title
    })# With Progress Function End
  })
  

  
  lapply(1:10, function(i) {
    

    output[[paste0('b', i)]] <- renderUI({
      newsDF <- getNews(input$text,10)
      newsTitleLink <- newsDF[,1:2]
      tagList(paste0(i,") "), a(  as.character(newsTitleLink[i,1] ) ,   href=as.character(newsTitleLink[i,2]) ) ) 
    })
    
    
  })  
  
  output$ending <- renderText({
    withProgress(message = 'Done', value = 0, {
      incProgress(10/10, detail = paste("Done"))
      title <- paste("------------------------")
      title
    })# With Progress Function End
  })
  
  
  # Helper methods
      
  createStockData <- function(inputStockName) {

    stockData <- getSymbols(inputStockName,auto.assign = FALSE)
    
    stockDF <- as.data.frame(stockData)
    
    return(stockDF)
  }
  
  plotGraph <- function(inputDF) {
    
    ggplot(inputDF, aes(x=as.Date(date, "%Y-%m-%d"), y=var)) + geom_line()
    
  }
  
  plotMultiGraph <- function(inputDF_1,inputDF_2=NULL,inputDF_3=NULL,inputDF_4=NULL) {
    
    ggplot(data=inputDF_1, mapping= aes( x = as.Date(date, "%Y-%m-%d"), y=var )) + geom_line(
      data=inputDF_1) + geom_line(data=inputDF_2) + geom_line(data=inputDF_3) + geom_line(data=inputDF_4)
  }
  
  ## Accepts input of character vectors with stock symbols
  ## E.g. "APPL" for one stock, or c("APPL","GOOG","UAL") for multiple
  ## Returns a data frame with the metrics
  getMetrics <- function(inputStockName) {
    
    what_metrics <- yahooQF(c("Price/Sales", 
                              "P/E Ratio",
                              "Price/EPS Estimate Next Year",
                              "PEG Ratio",
                              "Dividend Yield", 
                              "Market Capitalization"))
    
    tickers <- c(inputStockName)
    
    # Get the metrics
    # Not all the metrics are returned by Yahoo.
    metrics <- getQuote(paste(tickers, sep="", collapse=";"), what=what_metrics)
    
    #Add tickers as the first column and remove the first column which had date stamps
    metrics <- data.frame(Symbol=tickers, metrics[,2:length(metrics)]) 
    
    #Change colnames
    colnames(metrics) <- c("Symbol", "Price-to-Sales", "Price-to-Earnings", 
                           "P/E Ratio (Forward Est. 1Y)", "Price-to-Earnings-Growth", "Div Yield", "Market Cap")
    
    return(metrics)    
  }
  
  drawWordCloud <- function(text) {
    
    # Load the data as a corpus
    docs <- Corpus(VectorSource(text))
    
    toSpace <- content_transformer(function (x , pattern ) gsub(pattern, " ", x))
    docs <- tm_map(docs, toSpace, "/")
    docs <- tm_map(docs, toSpace, "@")
    docs <- tm_map(docs, toSpace, "\\|")
    
    # Convert the text to lower case
    docs <- tm_map(docs, content_transformer(tolower))
    # Remove numbers
    docs <- tm_map(docs, removeNumbers)
    # Remove english common stopwords
    docs <- tm_map(docs, removeWords, stopwords("english"))
    # Remove your own stop word
    # specify your stopwords as a character vector
    docs <- tm_map(docs, removeWords, c("blabla1", "blabla2")) 
    # Remove punctuations
    docs <- tm_map(docs, removePunctuation)
    # Eliminate extra white spaces
    docs <- tm_map(docs, stripWhitespace)
    
    dtm <- TermDocumentMatrix(docs)
    m <- as.matrix(dtm)
    v <- sort(rowSums(m),decreasing=TRUE)
    d <- data.frame(word = names(v),freq=v)
    
    d <- d[c(-1:-5),]
    
    head(d, 10)
    
    set.seed(1234)
    wordcloud(words = d$word, freq = d$freq, min.freq = 1,
              max.words=300, random.order=FALSE, rot.per=0.35, 
              colors=brewer.pal(8, "Dark2"),scale = c(4,1))
    
  }

  
  ## Symbol is stock symbol e.g GOOG
  ## number is the number of stories to download, e.g. 50 will give you 50 stories
  getNews <- function(symbol, number){
    
    
    # construct url to news feed rss and encode it correctly
    url.b1 = 'http://www.google.com/finance/company_news?q='
    url    = paste(url.b1, symbol, '&output=rss', "&start=", 1,
                   "&num=", number, sep = '')
    url    = URLencode(url)
    
    # parse xml tree, get item nodes, extract data and return data frame
    doc   = xmlTreeParse(url, useInternalNodes = T);
    nodes = getNodeSet(doc, "//item");
    mydf  = ldply(nodes, as.data.frame(xmlToList))
    
    # clean up names of data frame
    names(mydf) = str_replace_all(names(mydf), "value\\.", "")
    
    # convert pubDate to date-time object and convert time zone
    mydf$pubDate = strptime(mydf$pubDate, 
                            format = '%a, %d %b %Y %H:%M:%S', tz = 'GMT')
    mydf$pubDate = with_tz(mydf$pubDate, tz = 'America/New_york')
    
    # drop guid.text and guid..attrs
    mydf$guid.text = mydf$guid..attrs = NULL
    
    
    mydf$title <- as.character(mydf$title)
    mydf$link <- as.character(mydf$link)
    mydf$description <- as.character(mydf$description)
    
    ## Remove HTML tags
    regex_html <- "</?\\w+((\\s+\\w+(\\s*=\\s*(?:\".*?\"|'.*?'|[\\^'\">\\s]+))?)+\\s*|\\s*)/?>"
    
    mydf$description <- gsub("\n","",mydf$description)
    
    mydf$description <- gsub(regex_html,"",mydf$description,perl = TRUE)

    return(mydf)    
  }
  
  
  ## symbol e.g. GOOG, number e.g. 50 for 50 stories to download
  # outputFilePath is the location to place the file
  createNewsTXT <- function(symbol,number,outputFilePath="stockNews.txt") {
    
    stockNewsDF <- getNews(symbol,number)
    
    titles <- paste(stockNewsDF$title,collapse = ",")
    
    descriptions <- paste(stockNewsDF$description,collapse = ",")
    
    bigText <- paste(titles,descriptions,sep=" ")
    
    write(bigText,outputFilePath)
    
  }
  
})

final-project/metricsAPI.R

#Load necessary libraries

require(quantmod)

## Accepts input of character vectors with stock symbols
## E.g. "APPL" for one stock, or c("APPL","GOOG","UAL") for multiple
## Returns a data frame with the metrics
getMetrics <- function(inputStockName) {
  
  what_metrics <- yahooQF(c("Price/Sales", 
                            "P/E Ratio",
                            "Price/EPS Estimate Next Year",
                            "PEG Ratio",
                            "Dividend Yield", 
                            "Market Capitalization"))
  
  tickers <- c(inputStockName)
  
  # Get the metrics
  # Not all the metrics are returned by Yahoo.
  metrics <- getQuote(paste(tickers, sep="", collapse=";"), what=what_metrics)
  
  #Add tickers as the first column and remove the first column which had date stamps
  metrics <- data.frame(Symbol=tickers, metrics[,2:length(metrics)]) 
  
  #Change colnames
  colnames(metrics) <- c("Symbol", "Revenue Multiple", "Earnings Multiple", 
                         "Earnings Multiple (Forward)", "Price-to-Earnings-Growth", "Div Yield", "Market Cap")
  
  return(metrics)    
}

## Outputs a data frame with metrics data
metricDF <- getMetrics(c("AAPL","GOOG"))

metricDF

# OR Single Stock

metricDF2 <- getMetrics("AAPL")

metricDF2





final-project/stockNewsAPI.R
# load libraries
require(XML)
require(plyr)
require(stringr)
require(lubridate)

## Symbol is stock symbol e.g GOOG
## number is the number of stories to download, e.g. 50 will give you 50 stories
getNews <- function(symbol, number){
  
  
  # construct url to news feed rss and encode it correctly
  url.b1 = 'http://www.google.com/finance/company_news?q='
  url    = paste(url.b1, symbol, '&output=rss', "&start=", 1,
                 "&num=", number, sep = '')
  url    = URLencode(url)
  
  # parse xml tree, get item nodes, extract data and return data frame
  doc   = xmlTreeParse(url, useInternalNodes = T);
  nodes = getNodeSet(doc, "//item");
  mydf  = ldply(nodes, as.data.frame(xmlToList))
  
  # clean up names of data frame
  names(mydf) = str_replace_all(names(mydf), "value\\.", "")
  
  # convert pubDate to date-time object and convert time zone
  mydf$pubDate = strptime(mydf$pubDate, 
                          format = '%a, %d %b %Y %H:%M:%S', tz = 'GMT')
  mydf$pubDate = with_tz(mydf$pubDate, tz = 'America/New_york')
  
  # drop guid.text and guid..attrs
  mydf$guid.text = mydf$guid..attrs = NULL
  
  
  mydf$title <- as.character(mydf$title)
  mydf$link <- as.character(mydf$link)
  mydf$description <- as.character(mydf$description)
  
  ## Remove HTML tags
  regex_html <- "</?\\w+((\\s+\\w+(\\s*=\\s*(?:\".*?\"|'.*?'|[\\^'\">\\s]+))?)+\\s*|\\s*)/?>"
  
  mydf$description <- gsub("\n","",mydf$description)
  
  mydf$description <- gsub(regex_2,"",mydf$description,perl = TRUE)
  
  
  
  return(mydf)    
}


## symbol e.g. GOOG, number e.g. 50 for 50 stories to download
# outputFilePath is the location to place the file
createNewsTXT <- function(symbol,number,outputFilePath="stockNews.txt") {
  
  stockNewsDF <- getNews(symbol,number)
  
  titles <- paste(stockNewsDF$title,collapse = ",")
  
  descriptions <- paste(stockNewsDF$description,collapse = ",")
  
  bigText <- paste(titles,descriptions,sep=" ")
  
  write(bigText,outputFilePath)
  
}

final-project/ggplotCombine.R
##### Combining GGPlots ####

# library(ggplot2)
# library(fImport)

### Note! New createStockData function based on quantmod package!
## outputs more data (10 years worth etc)

# Quantmod version instead of fImport, 
createStockData <- function(inputStockName) {
  
  stockData <- getSymbols(inputStockName,auto.assign = FALSE)
  
  stockDF <- as.data.frame(stockData)
  
  return(stockDF)
}

## Helper Function

createStockData <- function(inputStockName) {
  stockTimeSeries = yahooSeries(inputStockName)
  
  stockDF = as.data.frame(stockTimeSeries)
  
  return(stockDF)
  
}

plotGraph <- function(inputDF) {
  
  ggplot(inputDF, aes(x=as.Date(date, "%Y-%m-%d"), y=var)) + geom_line()
  
  
  
  
}

plotMultiGraph <- function(inputDF_1,inputDF_2=NULL,inputDF_3=NULL,inputDF_4=NULL) {
  
  ggplot(data=inputDF_1, mapping= aes( x = as.Date(date, "%Y-%m-%d"), y=var )) + geom_line(
    data=inputDF_1) + geom_line(data=inputDF_2) + geom_line(data=inputDF_3) + geom_line(data=inputDF_4)
}

## Test Code

GOOGStock <- createStockData("GOOG")
UALStock <- createStockData("UAL")
YHOOStock <- createStockData("YHOO")

# Note, maybe have only one data array, in case of multiple and causes bug?, potential api wrong data problem
dataGOOG <- data.frame(date = row.names(GOOGStock), var = GOOGStock[["GOOG.Adj.Close"]])
dataUAL <- data.frame(date = row.names(UALStock), var = UALStock[["UAL.Adj.Close"]])
dataYHOO <- data.frame(date = row.names(YHOOStock), var = YHOOStock[["YHOO.Adj.Close"]])


plotGraph(dataUAL)
plotGraph(dataYHOO)
plotGraph(dataGOOG)

plotMultiGraph(dataUAL,dataGOOG)
plotMultiGraph(dataYHOO,dataGOOG)
plotMultiGraph(dataYHOO,dataUAL)

# Able to call 3 data frames while the fourth data frame is null, can be expanded as necessary easily
plotMultiGraph(dataYHOO,dataUAL,dataGOOG)

# as.Date(date, "%Y-%m-%d")
GOOGStock[which(as.Date(row.names(GOOGStock), "%Y-%m-%d") > (max(as.Date(row.names(GOOGStock), "%Y-%m-%d")) - 9)),]





final-project/ui.R
#
# This is the user-interface definition of a Shiny web application. You can
# run the application by clicking 'Run App' above.
#
# Find out more about building applications with Shiny here:
# 
#    http://shiny.rstudio.com/
#

library(shiny)

shinyUI(pageWithSidebar(
  
  # Application title
  headerPanel("Stocks Visualizer"),
  
  sidebarPanel(
    selectInput("variable", "Variable:",
                list("Adj Close" = "Adjusted",
                     "Open" = "Open",
                     "High" = "High",
                     "Low" = "Low",
                     "Close" = "Close",
                     "Volume" = "Volume")),
    
    selectInput("select", "Time Frame",
                list("5 Day" = "last 5 days",
                     "1 Month" = "last 30 days",
                     "3 Months" = "last 3 month",
                     "1 Year" = "last 1 year",
                     "3 Years" = "last 3 year",
                     "5 Years" = "last 5 year"), selected = "last 1 year"),
    
    # checkboxInput("outliers", "Show outliers", FALSE),
    
    checkboxGroupInput("TA", "Technical Indicators:",
                       choiceNames =
                         list("Simple Moving Average","Bollinger Bands"),
                       choiceValues =
                         list("addSMA();", "addBBands();")
    ),
    textOutput("txt"),
    
    textInput('text', 'Stock Symbol:', 'GOOG'),
    submitButton("Submit")
    # actionButton('add', 'Add to List')
    # verbatimTextOutput
 
  ),
  
  mainPanel(
    h3(textOutput("caption")),
    
    plotOutput("stockPlot", height = 750, width = 850),
    
    br(),
    
    tableOutput("metrics"),
    
    br(),
    
    plotOutput("wordCloud", height = 750, width = 1000),
    
    h4(textOutput("newsTitle")),
    
    lapply(1:10, function(i) {
      uiOutput(paste0('b', i))
    }),
    
    h5(textOutput("ending"))
    
    
  )
  
))
